### PR TITLE
plugins/lsp: use a no-default option when there is no default provided

### PR DESCRIPTION
**Pull request by MattSturgeon**
As per https://github.com/nix-community/nixvim/pull/2365#discussion_r1793846911

### CODE REVIEW

**Review by khaneliman — APPROVED** [review]

File: plugins/lsp/language-servers/default.nix [hunk 1]
```diff
@@ -179,18 +179,6 @@ let
...
   lspPackages = import ../lsp-packages.nix;
 
-  getLspPackage =
-    name:
-    if lib.hasAttr name lspPackages.packages then
-      { package = lspPackages.packages.${name}; }
-    else if lib.hasAttr name lspPackages.customCmd then
-      { inherit (lspPackages.customCmd.${name}) package cmd; }
-    else
-      assert lib.assertMsg (lib.elem name lspPackages.unpackaged) "LSP server ${name} is unknown";
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/default.nix:189]: We should probably wait for #2401 before dropping this assertion.

File: plugins/lsp/language-servers/_mk-lsp.nix [hunk 2]
```diff
@@ -50,10 +50,18 @@ in
       package =
         if lib.isOption package then
           package
-        else
+        else if args ? package then
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/_mk-lsp.nix:53]:
```suggestion
        else if args ? ${package} then
```

Comment (MattSturgeon) [reply plugins/lsp/language-servers/_mk-lsp.nix:53]: (nevermind, I'm dumb... It was correct before :joy:)

**Review by traxys — APPROVED** [review]

**Comment by khaneliman** [comment]
Hmm eval failing with those accessed but not defined errors

**Comment by traxys** [comment]
Ah this requires adaptations in the tests to pass null to the package it seems like

File: plugins/lsp/language-servers/_mk-lsp.nix [hunk 3]
```diff
@@ -50,16 +50,32 @@ in
...
-          if (cfg.package or null) != null then if builtins.isFunction cmd then cmd cfg else cmd else null;
+          # TODO: do we really only want the default `cmd` when `package` is non-null?
+          if !(opt.package.isDefined or false) then
+            null
+          else if cfg.package == null then
+            null
+          else if builtins.isFunction cmd then
+            cmd cfg
+          else
+            cmd;
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/_mk-lsp.nix:78]: I don't want to address this in this PR, but I've added a TODO comment regardless.

Essentially, `plugins.lsp.servers.<name>.cmd` is eventually passed through to `plugins.lsp.enabledServers.*.extraOptions`, so in a way it is a freeform sub-option.

It also seems that whether `cfg.package` is null or not is unrelated to whether we should use the default cmd.

Perhaps this comes from our history of defining a command based on things like `lib.getExe cfg.package`? Hopefully we've moved away from doing that and can enforce a policy where default `cmd` values do not reference the nix-store and are always defined.

If we keep the current behavior, we should also have a `defaultText` that explains how the default is conditional.

**Comment by MattSturgeon** [comment]
> Hmm eval failing with those accessed but not defined errors

> Ah this requires adaptations in the tests to pass null to the package it seems like

Should all be resolved now @khaneliman @traxys

**Review by khaneliman — APPROVED** [review]

**Comment by MattSturgeon** [comment]
@mergify queue

**Pull request merged**